File: 2024/Day05/day5.py
#!/usr/bin/env python3
import sys
from collections import defaultdict

rules = []
pages = []
with open(sys.argv[1], 'r') as file:
    flag = True
    for line in file:

        if line == '\n':
            flag = not flag
            continue

        if flag:
            rules.append(line.rstrip())
        else:
            pages.append(line.rstrip())

new_rules = defaultdict(list)
_ = [new_rules[int(k)].append(int(i)) for rule in rules for k, i in [rule.split('|')]]
pages = [tuple(map(int,page.split(','))) for page in pages]

def reorder_page_list(x: (int)):
    x = list(x)
    for i, page in enumerate(x):
        for j, p in enumerate(x[:i]):
            if p in new_rules[page]:
                x[i], x[j] = x[j], x[i]
    
    return tuple(x)

cnt1, cnt2 = 0, 0
for page_list in pages:
    flag = True
    for i, page in enumerate(page_list):
        if any(p in new_rules[page] for p in page_list[:i]):
            flag = False
            page_list = reorder_page_list(page_list)
            break

    if flag:
        cnt1 += page_list[int(len(page_list)/2)] 

    if not flag:
        cnt2 += page_list[int(len(page_list)/2)] 

print(f'Part 1: {cnt1}')
print(f'Part 2: {cnt2}')

File: 2024/Day05/day5.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <sstream>
#include <string>
#include <map>
#include <algorithm>

using namespace std;

vector<string> split(const string& line, const string& delimeter);
vector<int> convertToInt(const vector<string>& stringNumbers);
void reorder_page_list(vector<int>& x, map<int, vector<int> > rules);

// 'in' function: checks if an element exists in a vector
template <typename T>
bool in(const T& element, const std::vector<T>& container) {
    return std::find(container.begin(), container.end(), element) != container.end();
}

// 'any' function: checks if any element in the vector satisfies a condition
template <typename T, typename Func>
bool any(const std::vector<T>& container, Func condition) {
    return std::any_of(container.begin(), container.end(), condition);
}

int main() {
    ifstream f("input.txt");
    if (!f.is_open()) {
        cerr << "Error: Unable to open file." << endl;
        return 1;
    }

    map<int, vector<int> > rules;
    vector<vector<int> > page_lists;
    string line;
    bool flag = true;
    int i;
    vector<string> ptmp;
    while (getline(f, line)) {
        if (line.empty()) {
            flag = false;
            continue;
        }
        if (flag) {
            i = line.find('|');
            rules[stoi(line.substr(0,i))].push_back(stoi(line.substr(i+1,line.size())));
        } else {
            ptmp = split(line,",");  
            page_lists.push_back(convertToInt(ptmp));
        }
        
    }

    int tot1 = 0, tot2 = 0; 

    for (auto& page_list : page_lists) {
        bool flag = true; 
        for (int i = 0; i < page_list.size(); i++) {
            if (any(std::vector<int>(page_list.begin(), page_list.begin() + i), 
                    [&](int p) { return in(p, rules[page_list[i]]); })) {
                flag = false;
                reorder_page_list(page_list, rules);  
                break;
            }
        }

        if (flag) {
            tot1 += page_list[page_list.size() / 2]; 
        } else {
            tot2 += page_list[page_list.size() / 2];  
        }
    }

    // cout << "Rules:" << endl;
    // for (const auto& rule : rules) {
    //     cout << "Key: " << rule.first << " => Values: ";
    //     for (int val : rule.second) {
    //         cout << val << " ";
    //     }
    //     cout << endl;
    // }

    // cout << "Pages:" << endl;
    // for (auto pages: page_lists) {
    //     for (int page: pages) {
    //         cout << page << " ";
    //     }
    //     cout << endl;
    // }

    cout << "Part 1: " << tot1 << endl;
    cout << "Part 2: " << tot2 << endl;
}


vector<string> split(const string& line, const string& delimiter) {
    vector<string> tokens;
    size_t start = 0, end = line.find(delimiter);

    while (end != string::npos) {
        tokens.push_back(line.substr(start, end - start));
        start = end + delimiter.length();
        end = line.find(delimiter, start);
    }

    tokens.push_back(line.substr(start));
    return tokens;
}

vector<int> convertToInt(const vector<string>& stringNumbers) {
    vector<int> intNumbers;
    transform(stringNumbers.begin(), stringNumbers.end(), back_inserter(intNumbers),
              [](const string& str) { return stoi(str); });
    return intNumbers;
}

void reorder_page_list(vector<int>& x, map<int, vector<int> > rules) {
    for (int i = 0; i < x.size(); i++) {
        for (int j = 0; j < i; j++) {
            for (int r = 0; r < rules.at(x[i]).size(); r++) {
                if (x[j] == rules.at(x[i])[r]) {
                    std::swap(x[i], x[j]);
                }
            }
        }
    }
}